File: C/7.c
/*
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13,
we can see that the 6th prime is 13.

What is the 10001st prime number?
*/

#include <math.h>
#include <stdio.h>
const int maxPrime = 200000;
int primesToFind=10001;
char x[maxPrime];
int i;
int main() {
    // This is an implementation of the Sieve of Eratosthenes
    // I initially set all the numbers to be possibly prime
	for(i=2;i<maxPrime;i++)
		x[i]=1;

	for(i=2;primesToFind;){
		if(x[++i]){
			for(int j=2*i;j<maxPrime;j+=i)
				x[j]=0;
            primesToFind--;
        }
    }
	printf("%d\n",i);	// should be: 104743
}

File: Python/7.py
"""
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13,
we can see that the 6th prime is 13.

What is the 10001st prime number?
"""

import math

maxPrime = 200000
primeToFind = 10001
x = [True for i in range(maxPrime)]
x[0] = x[1] = False  # manually set 0 and 1 to not be primes
for i in range(2, int(math.sqrt(maxPrime))):
    if x[i]:
        for j in range(i, math.ceil(maxPrime / i)):
            x[i * j] = False
primesFound = 0
for i in range(maxPrime):
    if x[i]:
        primesFound += 1
        # print(i)
        if primesFound == primeToFind:
            print(i)  # should be: 104743
            break
